src/plugins/intel_gpu/src/runtime/kernels_cache.cpp

```diff
@@ -131,15 +131,68 @@ void kernels_cache::get_program_source(const kernels_code& kernels_source_code,
                 current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
             }
 
-            // Create new kernels batch when the limit is reached
-            // and current kernel's entry_point is duplicated in this kernels batch
-            if (current_bucket.back().kernels_counter >= get_max_kernels_per_batch()
-                || current_bucket.back().entry_point_to_id.find(entry_point) != current_bucket.back().entry_point_to_id.end()) {
-                const auto& batch_id = static_cast<int32_t>(current_bucket.size());
-                current_bucket.push_back(batch_program(bucket_id, batch_id, options, batch_header_str));
-            }
+            auto get_base_kernel_name = [](std::string unique_kernel_name) -> std::string {
+                int matched = 0;
+                size_t pos = unique_kernel_name.length();
+                if (unique_kernel_name.substr(pos-4, 4).compare("__sa") == 0) {
+                    pos -= 4;
+                }
+                while (pos > 0) {
+                    pos -= 1;
+                    if (unique_kernel_name.at(pos) == '_') {
+                        matched += 1;
+                    }
+                    if (matched == 2) {
+                        break;
+                    }
+                }
+                return unique_kernel_name.substr(0, pos);
+            };
+
+            auto get_target_batch = [&]() -> batch_program& {
```

Comment:
Does the issue really happen due to multiple instances of same kernel in the batch or it's just related to batch size? As I remember, if program source is too large, then igc may produce worse binary

Reply:
It's not clear what the root cause is as of now. BTW, it looks like that it is not due to program source size. In my test, the issue was gone when I commented out just one line.